### ISSUE TITLE
Typescript Type Definitions

### ISSUE THREAD

**Issue by wwahammy** [OpenEVSE/node-openevse#11]
Hey all, I just stumbled on this after looking at the OpenEVSE wifi gui.  I was looking at creating my own version of the Wifi GUI and I was going to use this library for it. That said, I was planning on using Typescript since I find it's very convenient for reliability. To do so, I'd need to create type definitions. I don't mind doing that but I wanted to know a few things before doing so:

* Would you like the types included in the project or have me put them on DefinitelyTyped? It's usually ideal to keep them in the main project but that's just to keep everything consistent.
* Is the API of this library likely to stay stable? Or if not, what is the process for changing it? I'm happy to update the types as needed but I want to make sure things stay consistent between the types and the actual API.

I'm really excited to start working with this; I really love my OpenEVSE :)

**Comment by jeremypoulter** [comment]
This library was mainly for a Linux/Node port of the OpenEVSE WiFi module, that project has been put on hold and thus so has this. Also be aware that RAPI comme over HTTP are being deprecated so this library may be of less use.

That being said if you wish to still make a contribution I am happy to support you. Adding types would be great and the API is ok to change.

**Comment by wwahammy** [comment]
@jeremypoulter ah alrighty. So what is the best process for building my own OpenEVSE wifi gui? https://github.com/OpenEVSE/openevse_wifi_gui is I think where it happens and one of the file said it was looking to reference this sooo... it's not clear to me where I can start :)

**Comment by jeremypoulter** [comment]
Unfortunately it is going to be a case of looking at the existing GUI, don't really have any docs on it yet. I am intending on creating an OpenAPI doc for the API at some point but unfortunately not high on the list. But feel free to ask questions

**Comment by jeremypoulter** [comment]
@wwahammy what are your goals for your UI? I have been thinking the current UI probably needs a redesign, been thinking doing a Material-UI/React based UI that properly uses webpack would be great

**Comment by wwahammy** [comment]
@jeremypoulter UI-wise, you've almost exactly defined what I want to do. I want to get better with React and Material-UI and making the Webpack build work "right" would be part of that.

The original motivating idea was that I'd like to add [WebThing](https://webthings.io/) support to OpenEVSE. I had originally thought it could be done through server-side Node on the charger but if I'm understanding things correctly, the server side software is compiled C code. Is that accurate?

**Comment by jeremypoulter** [comment]
Correct, the server side is C++, although we do have a mostly complete Node JS version from when we were looking at a replacement for the ESP8266.

I am having to look again at improving the webpack config as I need to use a node JS library on the front-end, but certainly a new UI would be great.

On the WedThings front there looks to be a Arduino library for the ESP32 so that is a good start.

**Comment by wwahammy** [comment]
@jeremypoulter I'm getting started on this task but I had two questions:

First: is there a better place to have this discussion?

Second: if, as I understand it, this repo doesn't have the best implementation of a javascript API for OpenEVSE, where is the best implementation?

**Comment by jeremypoulter** [comment]
@wwahammy 

> First: is there a better place to have this discussion?

Probably the best place to discuss is on the main repo: https://github.com/OpenEVSE/ESP32_WiFi_V4.x

> Second: if, as I understand it, this repo doesn't have the best implementation of a javascript API for OpenEVSE, where is the best implementation?

There isn't really a standalone JS API right now, probably the best place to look at is the [GUI](https://github.com/OpenEVSE/openevse_wifi_gui) or the [server](https://github.com/OpenEVSE/ESP32_WiFi_V4.x/blob/master/src/web_server.cpp).
